pkg/epoll/epoll.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2020 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package epoll
+
+import (
+	"errors"
+	"os"
+	"syscall"
+)
+
+const (
+	EPOLLIN  EventType = syscall.EPOLLIN
+	EPOLLPRI EventType = syscall.EPOLLPRI
+	EPOLLERR EventType = syscall.EPOLLERR
+	EPOLLOUT EventType = syscall.EPOLLOUT
+	EPOLLHUP EventType = syscall.EPOLLHUP
+
+	timeout int = 10
+)
+
+var (
+	ErrFileAlreadyWatched error = errors.New("file being watched already")
+	ErrInvalidEventType   error = errors.New("invalid event type")
+	ErrWatcherNotFound    error = errors.New("wacther not found")
+)
+
+type Epoll struct {
+	watchers      []Watcher
+	fileToWatcher map[string]*Watcher
+	fdToWatcher   map[int32]*Watcher
+	eventChan     chan Event
+	epfd          int
+	active        bool
+}
+
+type Event struct {
+	fileName string
+	events   uint32
+}
+
+type EventType uint32
+type Watcher struct {
+	fd *os.File
+	// The complete path of the file to watch. The file name is also used to
+	// uniquely identify the watcher
+	FileName string
+	// The event types to watch for
+	EventTypes []EventType
+}
+
+func New() (Epoll, error) {
+	epfd, err := syscall.EpollCreate(1)
+	if err != nil {
+		return Epoll{}, err
+	}
+	return Epoll{
+		watchers:      make([]Watcher, 0, 2),
```

Comment:
Why is the size of 2 required here? Can we just create the array instead of specifying the size.

Reply:
oh, so the length of the slice is still 0. I specified a capacity of 2 because users of the package will at least add one watcher. so, just kept a non-zero capacity.

Reply:
but we are appending to watcher slice right. So even if its zero size, it will get updated. 0 size should be fine. Because any integer that you keep in code, you will have to provide suitable explanation why it is being used.

Reply:
will update